state/protocol/badger/mutator.go

```diff
@@ -425,8 +425,21 @@ func (m *FollowerState) insert(ctx context.Context, candidate *flow.Block, last
 	defer span.End()
 
 	blockID := candidate.ID()
+	parentID := candidate.Header.ParentID
 	latestSealID := last.ID()
 
+	parent, err := m.headers.ByBlockID(parentID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve block header for %x: %w", parentID, err)
+	}
+	// root blocks and blocks below the root block are considered as "processed",
+	// so we don't want to trigger `BlockProcessable` event for them.
+	var rootHeight uint64
+	err = m.db.View(operation.RetrieveRootHeight(&rootHeight))
+	if err != nil {
+		return fmt.Errorf("could not retrieve root block's height: %w", err)
+	}
```

Comment:
I think `rootHeight` is a constant, that should not change while the node is up, right? Therefore, I think we should run this query in the `FollowerState` constructor and cache the result. We can then remove the query here and also in the [`guaranteeExtend` method](https://github.com/onflow/flow-go/blob/2401a43f30735ffe1527befd46681895c4d70282/state/protocol/badger/mutator.go#L253)

Reply:
Yes, good point. Added a cache in https://github.com/onflow/flow-go/pull/3303/commits/7cfdba093264fd6b1e1016333cf0490147a1d4f6 https://github.com/onflow/flow-go/pull/3303/commits/fd59222414bfd27e0c48eddbb684713447eef3f4